### PR TITLE
First Enroll Pull Request

### PR DESCRIPTION
**Pull request by PengRen**
Enroll plugin, Ontolgy, UserModel, keyboard and etc.
## 

Hi, This is Morteza, this is a new Pull Request for Peng's latest commits, but generally, I believe his most recent commits (after initial pull request) should have been automatically included in the initial pull request if committed and synched properly.

### CODE REVIEW

File: core/Client/Client.UI/MainWindow.xaml.cs [hunk 1]
```diff
@@ -152,6 +152,7 @@ private void LayoutPageWithPlugin()
 
 		private void LayoutPageWithoutPlugin()
 		{
+            MainGrid.Children.Remove(PluginContainer);
```
Comment (charlesrich) [inline core/Client/Client.UI/MainWindow.xaml.cs:155]: What bug is this fixing and how does it fix it?

File: core/Client/SoftKeyboard.UI/Keyboard.xaml [hunk 2]
```diff
@@ -9,7 +9,9 @@
              d:DesignHeight="500" d:DesignWidth="1400">
     <StackPanel Name="verticalPanel">
         <StackPanel Name="row0" Orientation="Horizontal" HorizontalAlignment="Center">
+            <StackPanel Name="x" Orientation="Horizontal" HorizontalAlignment="Center">
```
Comment (charlesrich) [inline core/Client/SoftKeyboard.UI/Keyboard.xaml]: Fix indentation to reflect new level of StackPanel's here and below.

File: core/Client/SoftKeyboard.UI/Keyboard.xaml [hunk 3]
```diff
@@ -21,7 +23,9 @@
             <local:ShiftViewKey NormalContent="9" ShiftContent="(" Shift="{Binding Shift, ElementName=keyboard}" />
             <local:ShiftViewKey NormalContent="0" ShiftContent=")" Shift="{Binding Shift, ElementName=keyboard}" />
             <local:ShiftViewKey NormalContent="-" ShiftContent="_" Shift="{Binding Shift, ElementName=keyboard}" />
+            <StackPanel Name="y" Orientation="Horizontal" HorizontalAlignment="Center">
```
Comment (charlesrich) [inline core/Client/SoftKeyboard.UI/Keyboard.xaml]: Use more descriptive name than "y" (and don't forget to change in usage)

File: core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs [hunk 4]
```diff
@@ -22,14 +22,27 @@ namespace SoftKeyboard.UI
 	/// </summary>
 	public partial class KeyboardPanel : UserControl
 	{
+
+        private static bool isOverflow = false;
```
Comment (charlesrich) [inline core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs:25]: Fix indentation in all added lines in this file to correspond with original file (it may be a tab problem?)

File: core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs [hunk 5]
```diff
@@ -37,9 +50,28 @@ private static void keyTypedHandler(object sender, RoutedEventArgs e)
...
+            Typeface typeface = new Typeface(panel.textBox.FontFamily, panel.textBox.FontStyle, panel.textBox.FontWeight, panel.textBox.FontStretch);
+            FormattedText ft = new FormattedText(panel.textBox.Text, System.Globalization.CultureInfo.CurrentCulture, System.Windows.FlowDirection.LeftToRight, typeface, panel.textBox.FontSize, Brushes.Black);
+            if (ft.Width > panel.textBox.ActualWidth - 5)
+            {
+                isOverflow = true;
+                panel.textBox.Background = new SolidColorBrush(Colors.Red);
+                panel.textBox.IsReadOnly = true;
+                JObject data = new JObject();
+                data.Add("event", "exceed");
+                data.Add("prompt", "The input is too long, please use the red arrow key to shorten it.");
```
Comment (charlesrich) [inline core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs]: I assume this is the default prompt (which will usually be overridden by a plugin that uses the KB).  Since the red key is also for making corrections, change prompt to: "Enter text (use red arrow key to rub out):"

File: core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs [hunk 6]
```diff
@@ -37,9 +50,28 @@ private static void keyTypedHandler(object sender, RoutedEventArgs e)
...
+                data.Add("text", panel.textBox.Text);
+                panel.remote.Send("keyboard.exceed", data);
+                panel.contextLabel.Content = "Limit exceeded, please press the red arrow key.";
```
Comment (charlesrich) [inline core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs]: Better prompt:  "Too many characters (use red arrow key to rub out):"

File: core/Client/SoftKeyboard.UI/SoftKeyboardPlugin.cs [hunk 7]
```diff
@@ -34,7 +34,9 @@ public void Dispose()
 
         public Viewbox GetPluginContainer()
         {
-            return pluginContainer;
+            //keyboard
+            //TODO
```
Comment (charlesrich) [inline core/Client/SoftKeyboard.UI/SoftKeyboardPlugin.cs:38]: What does this mean?

File: core/agent/java/edu/wpi/always/client/Keyboard.java [hunk 8]
```diff
@@ -1,8 +1,12 @@
 package edu.wpi.always.client;
 
-public interface Keyboard {
+import java.io.IOException;
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/Keyboard.java]: Remove unused import?   Set your preferences to flag unused imports as warnings.  Also see Source > Organize Imports command in Eclipse.

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java [hunk 9]
```diff
@@ -1,42 +1,90 @@
 package edu.wpi.always.client;
 
+import edu.wpi.disco.Agent;
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java]: These look like unused imports from earlier experimentation.  See earlier comment about managing imports better.

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java [hunk 10]
```diff
@@ -1,42 +1,90 @@
...
+import edu.wpi.disco.Interaction;
+import edu.wpi.disco.lang.Say;
+import edu.wpi.disco.lang.Utterance;
 import edu.wpi.disco.rt.menu.*;
 
-public abstract class KeyboardAdjacencyPair<C> extends AdjacencyPairBase<C> {
+public abstract class KeyboardAdjacencyPair<C> extends AdjacencyPairBase<C>{
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java]: Always follow the formatting conventions of the file you are editing, e.g., leave the space before the {.   You can adjust your formatting preferences in Eclipse to match the AlwaysOn code base.

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java [hunk 11]
```diff
@@ -1,42 +1,90 @@
...
 
-   private String prompt;
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java:7]: Adjust indentation to match surrounding code here and below in this file.

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java [hunk 12]
```diff
@@ -1,42 +1,90 @@
...
+		this(prompt, null, keyboard);
+	}
 
-   public KeyboardAdjacencyPair (String prompt, C context, Keyboard keyboard) {
-      super("Let me know when you're done", context);
-      this.prompt = prompt;
-      this.keyboard = keyboard;
-      choice("I'm Done", new DialogStateTransition() {
+	public KeyboardAdjacencyPair (String prompt, C context, Keyboard keyboard) {
+		super(prompt + "Let me know when you're done", context);
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java]: Does it really make sense to append the prompt on the front of the message????

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java [hunk 13]
```diff
@@ -1,42 +1,90 @@
...
-      keyboard.showKeyboard(prompt);
-   }
+	public KeyboardAdjacencyPair (String prompt, String showText, C context, Keyboard keyboard) {
+		super(prompt + "Let me know when you're done", context);
+		this.prompt = showText;
+		this.keyboard = keyboard;
+		choice("I'm Done", new DialogStateTransition() {
 
-   public abstract AdjacencyPair success (String text);
+			@Override
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java:41]: Does this code work correctly yet?  If not, do NOT include it in your pull request at this time.  You can/should be selective in your pull request.

File: core/agent/java/edu/wpi/always/client/KeyboardMessageHandler.java [hunk 14]
```diff
@@ -2,37 +2,70 @@
 
 import com.google.gson.JsonObject;
 import edu.wpi.always.client.ClientPluginUtils.InstanceReuseMode;
+import edu.wpi.always.cm.ProposalBuilder;
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/KeyboardMessageHandler.java]: Organize Imports.

File: core/agent/java/edu/wpi/always/user/people/ValidPhoneAgeBirthday.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
+package edu.wpi.always.user.people;
+
+import java.util.regex.Pattern;
+
+public class ValidPhoneAgeBirthday {
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/user/people/ValidPhoneAgeBirthday.java:5]: Better style to add these static methods to Person class and remove this class.

File: core/agent/java/edu/wpi/always/user/places/ZipCodes.java [hunk 16]
```diff
@@ -6,69 +6,167 @@
...
+				String line = s.nextLine();
+				if(!line.isEmpty()) {
+					StateEntry stateEntry = new StateEntry(line);
+					stateData.put(stateEntry.getState().toLowerCase(), stateEntry);
+					stateAbbrevData.put(stateEntry.getStateAbbrev().get(0).toLowerCase(), stateEntry);
+				}
+			}
+		}
+		try (Scanner s = new Scanner(new BufferedInputStream(getClass()
+				.getResourceAsStream("/edu/wpi/always/user/places/ZipCodes.csv")))) {
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/user/places/ZipCodes.java]: Why does this read ZipCode.csv twice?

File: core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java [hunk 17]
```diff
@@ -1,14 +1,18 @@
 package edu.wpi.always.user.owl;
 
+import edu.wpi.always.user.UserModel;
 import edu.wpi.always.user.people.*;
 import edu.wpi.always.user.people.Person.Gender;
 import edu.wpi.always.user.people.Person.Relationship;
+import edu.wpi.always.user.places.Place;
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java:7]: Why is this import needed? Organize Imports.

File: core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java [hunk 18]
```diff
@@ -18,9 +22,15 @@ public OntologyPeopleManager (Ontology ontology) {
    }
 
    // MUST be called for getUser to work (called in OntologyUserModel constructor)
-   void setUserModel (OntologyUserModel model) {
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java:21]: For encapsulation reasons, I don't want to add a public setUserModel method to PeopleManager.  Please revert to original design; talk to Morteza or me if you have problems.

File: core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java [hunk 19]
```diff
@@ -48,13 +58,41 @@ public OntologyPerson addPerson (String name, Relationship relationship,
          person.setGender(gender);
       return person;
    }
+   
+	@Override
+	public OntologyPerson addPerson(String name, Relationship relationship, Gender gender,
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java:53]: To avoid code duplication, also change the 3-argument version of addPerson to call this one with nulls.

File: core/agent/java/edu/wpi/always/user/owl/OntologyPerson.java [hunk 20]
```diff
@@ -3,161 +3,216 @@
...
-         return true;
-      if ( obj == null )
-         return false;
-      if ( !(obj instanceof Person) )
-         return false;
-      Person other = (Person) obj;
-      return other.getName().equals(getName());
-   }
+	public static final String PERSON_CLASS = "Person";
+	public static final String USER_CLASS = "User";
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/user/owl/OntologyPerson.java:15]: Because you changed indentation inconsistently below, it makes it very hard to see what the actual change it.  Please follow indentation conventions of the file you are editing.   Fix your preferences in Eclipse if you have to.

File: user/Activities.xml [hunk 21]
```diff
@@ -15,6 +15,8 @@
   <task id="PlayRummy"/>
 
   <task id="DiscussWeather"/>
+  
+  <task id="EnrollUser"/>
```
Comment (charlesrich) [inline user/Activities.xml:19]: Please do not edit this file.  I will do that.

File: user/Activities.properties [hunk 22]
```diff
@@ -10,6 +10,7 @@
 UseCalendar@plugin = edu.wpi.always.calendar.CalendarPlugin
```
Comment (charlesrich) [inline user/Activities.properties:10]: Please do not edit this file.  I will do that.

File: plugins/weather-plugin/weather/java/edu/wpi/always/weather/wunderground/WundergroundJSON.java [hunk 23]
```diff
@@ -1,5 +1,7 @@
 package edu.wpi.always.weather.wunderground;
 
+import edu.wpi.always.user.UserModel;
```
Comment (charlesrich) [inline plugins/weather-plugin/weather/java/edu/wpi/always/weather/wunderground/WundergroundJSON.java]: Be selective in pull request.  This is not ready yet.

File: plugins/weather-plugin/weather/java/edu/wpi/always/weather/WeatherPlugin.java [hunk 24]
```diff
@@ -29,7 +29,7 @@ public static void main (String[] args) {
       // testing new user property extension (see Weather.owl)
```
Comment (charlesrich) [inline plugins/weather-plugin/weather/java/edu/wpi/always/weather/WeatherPlugin.java:29]: Don't do pull request on weather yet.

File: core/Client/Client.UI/MainWindow.xaml.cs (hunk 1 above)
Comment (PengRen) [inline core/Client/Client.UI/MainWindow.xaml.cs:155]: This is used to fix the keyboard overlapping bug. Must first remove the original child from MainGrid and then add a new ViewBox.

File: core/Client/SoftKeyboard.UI/SoftKeyboardPlugin.cs (hunk 7 above)
Comment (PengRen) [inline core/Client/SoftKeyboard.UI/SoftKeyboardPlugin.cs:38]: This is in there after one fetch and merge but I will fixed that.

File: core/agent/java/edu/wpi/always/client/Keyboard.java (hunk 8 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/client/Keyboard.java]: Thank you for pointing out that.

Tag: Fixed

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java (hunk 9 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java]: Fixed

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java (hunk 12 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java]: To let the agent speak out the prompt, it is necessary.

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java (hunk 13 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java:41]: This code works correctly. It is a version of constructor to let the agent speak and at the same time display two different prompts.

File: core/agent/java/edu/wpi/always/user/places/ZipCodes.java (hunk 16 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/user/places/ZipCodes.java]: Sorry about the duplication, I will fix it.

File: core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java (hunk 17 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java:7]: The method addPerson() has a parameter whose type is Place.

File: core/Client/SoftKeyboard.UI/Keyboard.xaml (hunk 2 above)
Comment (PengRen) [inline core/Client/SoftKeyboard.UI/Keyboard.xaml]: fixed

File: core/Client/SoftKeyboard.UI/Keyboard.xaml (hunk 3 above)
Comment (PengRen) [inline core/Client/SoftKeyboard.UI/Keyboard.xaml]: fixed

File: core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs (hunk 4 above)
Comment (PengRen) [inline core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs:25]: Okay, I will be more careful about the indentation.

File: core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs (hunk 6 above)
Comment (PengRen) [inline core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs]: fixed

File: core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs (hunk 5 above)
Comment (PengRen) [inline core/Client/SoftKeyboard.UI/KeyboardPanel.xaml.cs]: fixed

File: core/Client/SoftKeyboard.UI/SoftKeyboardPlugin.cs (hunk 7 above)
Comment (PengRen) [inline core/Client/SoftKeyboard.UI/SoftKeyboardPlugin.cs:38]: Do you mean what does the "isNumeric" mean? This is a flag to control if use the numeric keyboard or not

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java (hunk 10 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java]: fixed. Thank you for pointing that out.

File: core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java (hunk 11 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/client/KeyboardAdjacencyPair.java:7]: fixed

File: core/agent/java/edu/wpi/always/client/KeyboardMessageHandler.java (hunk 14 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/client/KeyboardMessageHandler.java]: fixed

File: core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java (hunk 18 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java:21]: fixed

File: core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java (hunk 19 above)
Comment (PengRen) [inline core/agent/java/edu/wpi/always/user/owl/OntologyPeopleManager.java:53]: fixed

File: user/Activities.xml (hunk 21 above)
Comment (PengRen) [inline user/Activities.xml:19]: If I don't edit the file, I cannot run Enroll to demo. Do you mean I don't do a pull request for the files.

File: plugins/weather-plugin/weather/java/edu/wpi/always/weather/wunderground/WundergroundJSON.java (hunk 23 above)
Comment (PengRen) [inline plugins/weather-plugin/weather/java/edu/wpi/always/weather/wunderground/WundergroundJSON.java]: Okay, thank you for point that out.
